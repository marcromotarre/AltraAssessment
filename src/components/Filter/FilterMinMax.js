import React from 'react';
import styled from 'styled-components';
import PropTypes from 'prop-types';
import EyeHideSVG from '../../svg/eye-hide.svg';
import EyeShowSVG from '../../svg/eye-show.svg';

const FilterMinMaxContainer = styled.section`
  width: 100%;
  display: grid;
  grid-template-columns: 20% 30% 20% 30%;
  grid-template-rows: 20px 26px 20px;
  row-gap: 2px;
  grid-template-areas:
    '. . . m-show-hide-icon'
    'm-title m-min-input m-to m-max-input'
    '. m-min-value-info . m-max-value-info';
  label {
    font-family: 'HelveticaNeue-Thin';
    color: #000;
  }
  * {
    justify-self: center;
    align-self: center;
  }

  input {
    height: 26px;
    width: 80%;
    background: white;
    border: 1px solid black;
    font-family: 'HelveticaNeue-Thin';
    color: #000;
    font-size: 14px;
    justify-self: end;
    padding-left: 10px;
  }
`;

const Title = styled.label`
  grid-area: m-title;
  font-size: 14px;
  justify-self: start;
`;

const MinInput = styled.input`
  grid-area: m-min-input;
`;

const MinValueInfo = styled.label`
  grid-area: m-min-value-info;
  font-size: 10px;
  justify-self: end;
  align-self: start;
`;

const To = styled.label`
  grid-area: m-to;
  font-size: 14px;
`;

const MaxInput = styled.input`
  grid-area: m-max-input;
`;

const MaxValueInfo = styled.label`
  grid-area: m-max-value-info;
  font-size: 10px;
  justify-self: end;
  align-self: start;
`;

const ShowHideButton = styled.button`
  grid-area: m-show-hide-icon;
  display: flex;
  justify-content: center;
  align-items: center;
`;

const FilterMinMax = ({
  title,
  minLimit,
  maxLimit,
  minInputValue,
  maxInputValue,
  showTo,
  minChanged,
  maxChanged,
  showToChanged,
}) => {
  const handleMinInputChange = (event) => {
    if (event.target.value.match(/^[0-9]*$/g)) {
      minChanged(event.target.value);
    }
  };
  const handleMaxInputChange = (event) => {
    if (event.target.value.match(/^[0-9]*$/g)) {
      maxChanged(event.target.value);
    }
  };

  const handleShowTo = () => {
    showToChanged(!showTo);
  };

  return (
    <FilterMinMaxContainer>
      <Title>{title}:</Title>
      <MinInput
        value={minInputValue}
        onChange={handleMinInputChange}
        data-testid={`${title}-filterminmax-min-input`}
      />
      {showTo && <MinValueInfo>min: {Math.floor(minLimit)}</MinValueInfo>}
      {showTo && <To>to</To>}
      {showTo && (
        <MaxInput
          value={maxInputValue}
          onChange={handleMaxInputChange}
          data-testid={`${title}-filterminmax-max-input`}
        />
      )}
      {showTo && <MaxValueInfo>max: {Math.ceil(maxLimit)}</MaxValueInfo>}
      <ShowHideButton
        onClick={handleShowTo}
        data-testid={`${title}-filterminmax-show-hide-button`}
      >
        {!showTo && <EyeShowSVG width={11.43}></EyeShowSVG>}
        {showTo && <EyeHideSVG width={11.43}></EyeHideSVG>}
      </ShowHideButton>
    </FilterMinMaxContainer>
  );
};

FilterMinMax.propTypes = {
  title: PropTypes.string.isRequired,
  minLimit: PropTypes.number.isRequired,
  maxLimit: PropTypes.number.isRequired,
  minInputValue: PropTypes.string.isRequired,
  maxInputValue: PropTypes.string.isRequired,
  showTo: PropTypes.bool,
  minChanged: PropTypes.func.isRequired,
  maxChanged: PropTypes.func.isRequired,
  showToChanged: PropTypes.func.isRequired,
};

FilterMinMax.defaultProps = {
  showTo: true,
};

export default FilterMinMax;
