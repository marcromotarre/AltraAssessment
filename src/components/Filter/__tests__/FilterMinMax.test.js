import React from 'react';
import renderer from 'react-test-renderer';
import { render, fireEvent } from '@testing-library/react';
import { shallow } from 'enzyme';
import FilterMinMax from '../FilterMinMax';

/** this component should:
 *  - render FilterMinMax with:
 *      - filter title
 *      - to label (only renders if showTo is true):
 *      - show hide button (allow to show hide calling external function)
 *      - min:
 *          - input value (allow change value calling external function)
 *                  - only changes with numbers
 *          - total information, round floor (only renders if showTo is true)
 *      - max (only renders if showTo is true):
 *          - input value, round ceil (allow change value calling external function)
 *                  - only changes with numbers
 *          - total information
 */

const props = {
  title: 'title',
  minLimit: 10.9,
  maxLimit: 11.2,
  minInputValue: '23',
  maxInputValue: '45',
  minChanged: jest.fn(),
  maxChanged: jest.fn(),
  showToChanged: jest.fn(),
};

const numericEvent = {
  target: {
    value: '34',
  },
};

const nonNumericEvent = {
  target: {
    value: '34e',
  },
};

describe('FilterMinMax Component tests', () => {
  it('FilterMinMax Snapshot', () => {
    const tree = renderer.create(<FilterMinMax {...props} />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('FilterMinMax renders title', () => {
    const wrapper = shallow(<FilterMinMax {...props} />);
    const title = wrapper.childAt(0);
    expect(title.type().displayName).toBe('styled.label');
    expect(title.text()).toBe(`${props.title}:`);
  });

  it('FilterMinMax renders min input', () => {
    const wrapper = shallow(<FilterMinMax {...props} />);
    const minInput = wrapper.childAt(1);
    expect(minInput.type().displayName).toBe('styled.input');
    expect(minInput.prop('value')).toBe(`${props.minInputValue}`);
  });

  it('props.minChanged is called when user write a number, min input value does not change', () => {
    const { getByTestId } = render(<FilterMinMax {...props} />);
    const minInput = getByTestId('title-filterminmax-min-input');
    expect(minInput.value).toBe(`${props.minInputValue}`);
    expect(props.minChanged).toHaveBeenCalledTimes(0);
    fireEvent.change(minInput, nonNumericEvent);
    expect(props.minChanged).toHaveBeenCalledTimes(0);
    fireEvent.change(minInput, numericEvent);
    expect(props.minChanged).toHaveBeenCalledWith(numericEvent.target.value);
    expect(props.minChanged).toHaveBeenCalledTimes(1);
    expect(minInput.value).toBe(`${props.minInputValue}`);
  });

  it('FilterMinMax renders MinValueInfo', () => {
    const wrapper = shallow(<FilterMinMax {...props} />);
    const minValueInfo = wrapper.childAt(2);
    expect(minValueInfo.type().displayName).toBe('styled.label');
    expect(minValueInfo.text()).toBe(`min: ${Math.floor(props.minLimit)}`);
  });

  it('FilterMinMax renders title', () => {
    const wrapper = shallow(<FilterMinMax {...props} />);
    const to = wrapper.childAt(3);
    expect(to.type().displayName).toBe('styled.label');
    expect(to.text()).toBe(`to`);
  });

  it('FilterMinMax renders max input', () => {
    const wrapper = shallow(<FilterMinMax {...props} />);
    const maxInput = wrapper.childAt(4);
    expect(maxInput.type().displayName).toBe('styled.input');
    expect(maxInput.prop('value')).toBe(`${props.maxInputValue}`);
  });

  it('props.maxChanged is called when user write a number, max input value does not change', () => {
    const { getByTestId } = render(<FilterMinMax {...props} />);
    const maxInput = getByTestId('title-filterminmax-max-input');
    expect(maxInput.value).toBe(`${props.maxInputValue}`);
    expect(props.maxChanged).toHaveBeenCalledTimes(0);
    fireEvent.change(maxInput, nonNumericEvent);
    expect(props.maxChanged).toHaveBeenCalledTimes(0);
    fireEvent.change(maxInput, numericEvent);
    expect(props.maxChanged).toHaveBeenCalledWith(numericEvent.target.value);
    expect(props.maxChanged).toHaveBeenCalledTimes(1);
    expect(maxInput.value).toBe(`${props.maxInputValue}`);
  });

  it('FilterMinMax renders MaxValueInfo', () => {
    const wrapper = shallow(<FilterMinMax {...props} />);
    const maxValueInfo = wrapper.childAt(5);
    expect(maxValueInfo.type().displayName).toBe('styled.label');
    expect(maxValueInfo.text()).toBe(`max: ${Math.ceil(props.maxLimit)}`);
  });

  it('props.showToChanged is called when user write click on show hide button', () => {
    const { getByTestId } = render(<FilterMinMax {...props} />);
    const showHideButton = getByTestId('title-filterminmax-show-hide-button');
    expect(props.showToChanged).toHaveBeenCalledTimes(0);
    fireEvent.click(showHideButton);
    expect(props.showToChanged).toHaveBeenCalledWith(false);
    expect(props.showToChanged).toHaveBeenCalledTimes(1);
  });
});
