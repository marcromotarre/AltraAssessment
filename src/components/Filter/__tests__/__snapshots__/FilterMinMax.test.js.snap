// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FilterMinMax Component tests FilterMinMax Snapshot 1`] = `
<section
  className="sc-AxjAm cUObhj"
>
  <label
    className="sc-AxirZ pTpRo"
  >
    title
    :
  </label>
  <input
    className="sc-AxiKw kJgCdA"
    data-testid="title-filterminmax-min-input"
    onChange={[Function]}
    value="23"
  />
  <label
    className="sc-AxhCb iAbugP"
  >
    min: 
    10
  </label>
  <label
    className="sc-AxhUy cUdIeG"
  >
    to
  </label>
  <input
    className="sc-AxgMl eFELnN"
    data-testid="title-filterminmax-max-input"
    onChange={[Function]}
    value="45"
  />
  <label
    className="sc-AxheI cUaqWs"
  >
    max: 
    12
  </label>
  <button
    className="sc-Axmtr fdTANV"
    data-testid="title-filterminmax-show-hide-button"
    onClick={[Function]}
  >
    <svg
      width={11.43}
    />
  </button>
</section>
`;
